File: server/engine/file_format.py
"""Direct port of src/lib/fileFormat.ts."""

import re

FILE_FORMATS = ["csv", "parquet", "json", "delta"]


def infer_file_format(file_name: str) -> str | None:
    match = re.search(r"\.([a-z0-9]+)$", file_name.strip().lower())
    if not match:
        return None
    ext = match.group(1)
    if ext in ("csv", "tsv"):
        return "csv"
    if ext in ("parquet", "pq"):
        return "parquet"
    if ext in ("json", "jsonl", "ndjson"):
        return "json"
    if ext == "delta":
        return "delta"
    return None


def build_file_path(location: str | None, file_name: str | None) -> str:
    loc = (location or "").strip()
    name = (file_name or "").strip()
    if not loc:
        return name
    if not name:
        return loc
    sep = "" if re.search(r"[/\\]$", loc) else "/"
    return f"{loc}{sep}{name}"

File: src/lib/fileFormat.ts
export type FileFormat = 'csv' | 'parquet' | 'json' | 'delta';

export const FILE_FORMATS: FileFormat[] = ['csv', 'parquet', 'json', 'delta'];

/** Infers the Spark file format from a file name's extension, or null when it can't be determined. */
export function inferFileFormat(fileName: string): FileFormat | null {
  const ext = fileName.trim().toLowerCase().match(/\.([a-z0-9]+)$/)?.[1];
  if (!ext) return null;
  if (ext === 'csv' || ext === 'tsv') return 'csv';
  if (ext === 'parquet' || ext === 'pq') return 'parquet';
  if (ext === 'json' || ext === 'jsonl' || ext === 'ndjson') return 'json';
  if (ext === 'delta') return 'delta';
  return null;
}

/** Joins a file location and file name into a single path, tolerating a missing/trailing separator. */
export function buildFilePath(location: string | undefined, fileName: string | undefined): string {
  const loc = (location ?? '').trim();
  const name = (fileName ?? '').trim();
  if (!loc) return name;
  if (!name) return loc;
  const sep = /[/\\]$/.test(loc) ? '' : '/';
  return `${loc}${sep}${name}`;
}
